File: ios/Runner/ChimeMeetingPlugin.swift
import Flutter
import UIKit
import AmazonChimeSDK
import AVFoundation

public class ChimeMeetingPlugin: NSObject, FlutterPlugin, AudioVideoObserver, VideoTileObserver, RealtimeObserver {
    private var meetingSession: MeetingSession?
    private var eventSink: FlutterEventSink?

    private var localTileId: Int?
    private var remoteTileId: Int?

    private var activeLocalView: ChimeVideoPlatformView?
    private var activeRemoteView: ChimeVideoPlatformView?

    private var localAttendeeId: String?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.danish.m2m/chime", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "com.danish.m2m/chime_events", binaryMessenger: registrar.messenger())

        let instance = ChimeMeetingPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)

        registrar.register(
            ChimeVideoPlatformViewFactory(plugin: instance),
            withId: "com.danish.m2m/chime_video"
        )
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "join":
            guard let args = call.arguments as? [String: Any],
                  let meetingMap = args["meeting"] as? [String: Any],
                  let attendeeMap = args["attendee"] as? [String: Any] else {
                result(FlutterError(code: "invalid_arguments", message: "Arguments were invalid", details: nil))
                return
            }
            let videoEnabled = args["videoEnabled"] as? Bool ?? false
            joinMeeting(meetingMap: meetingMap, attendeeMap: attendeeMap, videoEnabled: videoEnabled, result: result)
        case "leave":
            leaveMeeting()
            result(nil)
        case "setMuted":
            guard let args = call.arguments as? [String: Any],
                  let muted = args["muted"] as? Bool else {
                result(FlutterError(code: "invalid_arguments", message: "Muted parameter missing", details: nil))
                return
            }
            if muted {
                meetingSession?.audioVideo.realtimeLocalMute()
            } else {
                meetingSession?.audioVideo.realtimeLocalUnmute()
            }
            result(nil)
        case "setCameraEnabled":
            guard let args = call.arguments as? [String: Any],
                  let enabled = args["enabled"] as? Bool else {
                result(FlutterError(code: "invalid_arguments", message: "Enabled parameter missing", details: nil))
                return
            }
            if enabled {
                meetingSession?.audioVideo.startLocalVideo()
            } else {
                meetingSession?.audioVideo.stopLocalVideo()
            }
            result(nil)
        case "switchCamera":
            meetingSession?.audioVideo.switchCamera()
            applyLocalMirror()
            result(nil)
        case "setSpeakerEnabled":
            guard let args = call.arguments as? [String: Any],
                  let enabled = args["enabled"] as? Bool else {
                result(FlutterError(code: "invalid_arguments", message: "Enabled parameter missing", details: nil))
                return
            }
            setSpeaker(enabled: enabled)
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func joinMeeting(
        meetingMap: [String: Any],
        attendeeMap: [String: Any],
        videoEnabled: Bool,
        result: @escaping FlutterResult
    ) {
        leaveMeeting()

        guard let meetingId = meetingMap["MeetingId"] as? String,
              let mediaPlacementMap = meetingMap["MediaPlacement"] as? [String: Any],
              let audioHostUrl = mediaPlacementMap["AudioHostUrl"] as? String,
              let audioFallbackUrl = mediaPlacementMap["AudioFallbackUrl"] as? String,
              let signalingUrl = mediaPlacementMap["SignalingUrl"] as? String,
              let turnControlUrl = mediaPlacementMap["TurnControlUrl"] as? String else {
            result(FlutterError(code: "invalid_arguments", message: "Invalid meeting schema", details: nil))
            return
        }

        let externalMeetingId = meetingMap["ExternalMeetingId"] as? String ?? ""
        let eventIngestionUrl = mediaPlacementMap["EventIngestionUrl"] as? String

        guard let attendeeId = attendeeMap["AttendeeId"] as? String,
              let joinToken = attendeeMap["JoinToken"] as? String else {
            result(FlutterError(code: "invalid_arguments", message: "Invalid attendee schema", details: nil))
            return
        }
        let externalUserId = attendeeMap["ExternalUserId"] as? String ?? ""

        self.localAttendeeId = attendeeId

        let mediaPlacement = MediaPlacement(
            audioFallbackUrl: audioFallbackUrl,
            audioHostUrl: audioHostUrl,
            signalingUrl: signalingUrl,
            turnControlUrl: turnControlUrl,
            eventIngestionUrl: eventIngestionUrl
        )

        let meeting = Meeting(
            externalMeetingId: externalMeetingId,
            mediaPlacement: mediaPlacement,
            mediaRegion: meetingMap["MediaRegion"] as? String ?? "us-east-1",
            meetingId: meetingId
        )

        let attendee = Attendee(
            attendeeId: attendeeId,
            externalUserId: externalUserId,
            joinToken: joinToken
        )

        let configuration = MeetingSessionConfiguration(
            createMeetingResponse: CreateMeetingResponse(meeting: meeting),
            createAttendeeResponse: CreateAttendeeResponse(attendee: attendee)
        )

        let logger = ConsoleLogger(name: "ChimeSDK")
        let session = DefaultMeetingSession(configuration: configuration, logger: logger)
        self.meetingSession = session

        session.audioVideo.addAudioVideoObserver(observer: self)
        session.audioVideo.addVideoTileObserver(observer: self)
        session.audioVideo.addRealtimeObserver(observer: self)

        do {
            try session.audioVideo.start()
            // Enable RECEIVING remote video  the master switch for accepting inbound
            // video frames. updateVideoSourceSubscriptions only selects which sources to
            // subscribe to; it does not enable reception. Without this the peer's tile
            // never arrives and both sides sit on "Waiting for video" (audio + local
            // self-view still work, which was the exact symptom). Safe no-op for
            // audio-only calls.
            session.audioVideo.startRemoteVideo()
            if videoEnabled {
                try session.audioVideo.startLocalVideo()
            }
            result(nil)
        } catch {
            result(FlutterError(code: "join_failed", message: error.localizedDescription, details: nil))
        }
    }

    private func leaveMeeting() {
        if let session = meetingSession {
            session.audioVideo.stop()
            session.audioVideo.removeAudioVideoObserver(observer: self)
            session.audioVideo.removeVideoTileObserver(observer: self)
            session.audioVideo.removeRealtimeObserver(observer: self)
        }
        meetingSession = nil
        localTileId = nil
        remoteTileId = nil
        localAttendeeId = nil
    }

    private func setSpeaker(enabled: Bool) {
        let session = AVAudioSession.sharedInstance()
        do {
            try session.overrideOutputAudioPort(enabled ? .speaker : .none)
        } catch {
            NSLog("Failed to override audio port: %@", error.localizedDescription)
        }
    }

    /// Mirror the local self-view ONLY when the front camera is active, the way
    /// users expect a selfie preview to behave. Back camera is never mirrored.
    /// `mirror` is render-only on the local DefaultVideoRenderView, so the stream
    /// sent to the remote peer is unaffected  they always see us correctly
    /// oriented, and published/captured frames are not flipped.
    private func applyLocalMirror() {
        let isFront = meetingSession?.audioVideo.getActiveCamera()?.type == .videoFrontCamera
        activeLocalView?.getVideoRenderView().mirror = isFront
    }

    func registerVideoView(view: ChimeVideoPlatformView, isLocal: Bool) {
        DispatchQueue.main.async {
            if isLocal {
                self.activeLocalView = view
                if let tileId = self.localTileId {
                    self.meetingSession?.audioVideo.bindVideoView(videoView: view.getVideoRenderView(), tileId: tileId)
                }
                self.applyLocalMirror()
            } else {
                self.activeRemoteView = view
                if let tileId = self.remoteTileId {
                    self.meetingSession?.audioVideo.bindVideoView(videoView: view.getVideoRenderView(), tileId: tileId)
                }
            }
        }
    }

    func unregisterVideoView(view: ChimeVideoPlatformView) {
        DispatchQueue.main.async {
            if view === self.activeLocalView {
                self.activeLocalView = nil
                if let tileId = self.localTileId {
                    self.meetingSession?.audioVideo.unbindVideoView(tileId: tileId)
                }
            } else if view === self.activeRemoteView {
                self.activeRemoteView = nil
                if let tileId = self.remoteTileId {
                    self.meetingSession?.audioVideo.unbindVideoView(tileId: tileId)
                }
            }
        }
    }

    private func sendEvent(eventName: String) {
        DispatchQueue.main.async {
            self.eventSink?(["event": eventName])
        }
    }

    // MARK: - FlutterStreamHandler
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }

    // MARK: - AudioVideoObserver
    public func audioVideoDidStart(sessionStatus: MeetingSessionStatus) {
        sendEvent(eventName: "connected")
    }

    public func audioVideoDidStartConnecting(reconnecting: Bool) {
        sendEvent(eventName: reconnecting ? "reconnecting" : "connecting")
    }

    public func audioVideoDidStop(sessionStatus: MeetingSessionStatus) {
        sendEvent(eventName: "disconnected")
    }

    public func audioVideoDidCancel() {
        sendEvent(eventName: "disconnected")
    }

    public func connectionDidBecomePoor() {}
    public func connectionDidRecover() {
        sendEvent(eventName: "connected")
    }

    public func videoSessionDidStartConnecting() {}
    public func videoSessionDidStart(sessionStatus: MeetingSessionStatus) {}
    public func videoSessionDidStop(sessionStatus: MeetingSessionStatus) {}

    // Mirror of the Android fix: explicitly subscribe to remote video sources so
    // the remote tile is delivered. Harmless if the platform already
    // auto-subscribes  subscribing to an active source is idempotent.
    public func remoteVideoSourcesDidBecomeAvailable(sources: [RemoteVideoSource]) {
        guard let audioVideo = meetingSession?.audioVideo, !sources.isEmpty else { return }
        var configs: [RemoteVideoSource: VideoSubscriptionConfiguration] = [:]
        for source in sources {
            configs[source] = VideoSubscriptionConfiguration(priority: .highest, targetResolution: .high)
        }
        audioVideo.updateVideoSourceSubscriptions(addedOrUpdated: configs, removed: [])
    }

    public func remoteVideoSourcesDidBecomeUnavailable(sources: [RemoteVideoSource]) {
        guard let audioVideo = meetingSession?.audioVideo, !sources.isEmpty else { return }
        audioVideo.updateVideoSourceSubscriptions(addedOrUpdated: [:], removed: sources)
    }

    // MARK: - VideoTileObserver
    public func videoTileDidAdd(tileState: VideoTileState) {
        if tileState.isLocalTile {
            localTileId = tileState.tileId
            sendEvent(eventName: "localVideoEnabled")
            if let view = activeLocalView {
                meetingSession?.audioVideo.bindVideoView(videoView: view.getVideoRenderView(), tileId: tileState.tileId)
            }
            applyLocalMirror()
        } else {
            remoteTileId = tileState.tileId
            sendEvent(eventName: "remoteVideoEnabled")
            if let view = activeRemoteView {
                meetingSession?.audioVideo.bindVideoView(videoView: view.getVideoRenderView(), tileId: tileState.tileId)
            }
        }
    }

    public func videoTileDidRemove(tileState: VideoTileState) {
        if tileState.isLocalTile {
            localTileId = nil
            sendEvent(eventName: "localVideoDisabled")
            meetingSession?.audioVideo.unbindVideoView(tileId: tileState.tileId)
        } else {
            remoteTileId = nil
            sendEvent(eventName: "remoteVideoDisabled")
            meetingSession?.audioVideo.unbindVideoView(tileId: tileState.tileId)
        }
    }

    public func videoTileDidPause(tileState: VideoTileState) {}
    public func videoTileDidResume(tileState: VideoTileState) {}
    public func videoTileSizeDidChange(tileState: VideoTileState) {}

    // MARK: - RealtimeObserver
    public func attendeesDidJoin(attendeeInfo: [AttendeeInfo]) {
        for info in attendeeInfo {
            if info.attendeeId != localAttendeeId {
                sendEvent(eventName: "remoteJoined")
            }
        }
    }

    public func attendeesDidLeave(attendeeInfo: [AttendeeInfo]) {
        for info in attendeeInfo {
            if info.attendeeId != localAttendeeId {
                sendEvent(eventName: "remoteLeft")
            }
        }
    }

    public func attendeesDidDrop(attendeeInfo: [AttendeeInfo]) {
        for info in attendeeInfo {
            if info.attendeeId != localAttendeeId {
                sendEvent(eventName: "remoteLeft")
            }
        }
    }

    public func volumeDidChange(volumeUpdates: [VolumeUpdate]) {}
    public func signalStrengthDidChange(signalUpdates: [SignalUpdate]) {}
    public func attendeesDidMute(attendeeInfo: [AttendeeInfo]) {}
    public func attendeesDidUnmute(attendeeInfo: [AttendeeInfo]) {}
}

File: android/app/src/main/java/com/danish/m2m/ChimeMeetingPlugin.kt
package com.danish.m2m

import android.content.Context
import android.media.AudioManager
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.amazonaws.services.chime.sdk.meetings.audiovideo.AudioVideoFacade
import com.amazonaws.services.chime.sdk.meetings.audiovideo.AudioVideoObserver
import com.amazonaws.services.chime.sdk.meetings.audiovideo.SignalUpdate
import com.amazonaws.services.chime.sdk.meetings.audiovideo.VolumeUpdate
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoTileObserver
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoTileState
import com.amazonaws.services.chime.sdk.meetings.device.MediaDevice
import com.amazonaws.services.chime.sdk.meetings.device.MediaDeviceType
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.RemoteVideoSource
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoPriority
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoResolution
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoSubscriptionConfiguration
import com.amazonaws.services.chime.sdk.meetings.audiovideo.AttendeeInfo
import com.amazonaws.services.chime.sdk.meetings.realtime.RealtimeObserver
import com.amazonaws.services.chime.sdk.meetings.session.Attendee
import com.amazonaws.services.chime.sdk.meetings.session.CreateAttendeeResponse
import com.amazonaws.services.chime.sdk.meetings.session.CreateMeetingResponse
import com.amazonaws.services.chime.sdk.meetings.session.DefaultMeetingSession
import com.amazonaws.services.chime.sdk.meetings.session.Meeting
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSession
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSessionConfiguration
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSessionStatus
import com.amazonaws.services.chime.sdk.meetings.session.MediaPlacement
import com.amazonaws.services.chime.sdk.meetings.utils.logger.ConsoleLogger
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class ChimeMeetingPlugin(private val context: Context) :
    MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler,
    AudioVideoObserver,
    VideoTileObserver,
    RealtimeObserver {

    private var meetingSession: MeetingSession? = null
    private var eventSink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private val audioResetRunnable = Runnable { resetAndroidAudioAfterCall() }

    private var localTileId: Int? = null
    private var remoteTileId: Int? = null

    private var activeLocalView: ChimeVideoPlatformView? = null
    private var activeRemoteView: ChimeVideoPlatformView? = null

    private var localAttendeeId: String? = null
    private var desiredMuted: Boolean = false

    companion object {
        private const val TAG = "M2MChimeMeeting"
        private const val AUDIO_RESET_DELAY_MS = 500L

        fun registerWith(flutterEngine: FlutterEngine, context: Context) {
            val plugin = ChimeMeetingPlugin(context)

            MethodChannel(flutterEngine.dartExecutor.binaryMessenger, "com.danish.m2m/chime")
                .setMethodCallHandler(plugin)

            EventChannel(flutterEngine.dartExecutor.binaryMessenger, "com.danish.m2m/chime_events")
                .setStreamHandler(plugin)

            flutterEngine.platformViewsController.registry.registerViewFactory(
                "com.danish.m2m/chime_video",
                ChimeVideoPlatformViewFactory(plugin)
            )
        }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "join" -> {
                val meetingMap = call.argument<Map<String, Any>>("meeting")
                val attendeeMap = call.argument<Map<String, Any>>("attendee")
                val videoEnabled = call.argument<Boolean>("videoEnabled") ?: false

                if (meetingMap == null || attendeeMap == null) {
                    result.error("invalid_arguments", "Meeting or Attendee map was null", null)
                    return
                }

                try {
                    joinMeeting(meetingMap, attendeeMap, videoEnabled)
                    result.success(null)
                } catch (e: Exception) {
                    result.error("join_failed", e.message, null)
                }
            }
            "leave" -> {
                leaveMeeting()
                result.success(null)
            }
            "setMuted" -> {
                val muted = call.argument<Boolean>("muted") ?: false
                val audioVideo = meetingSession?.audioVideo
                if (audioVideo == null) {
                    result.error("no_active_meeting", "No active meeting audio session.", null)
                    return
                }
                desiredMuted = muted
                applyLocalMute(audioVideo, muted)
                result.success(null)
            }
            "setCameraEnabled" -> {
                val enabled = call.argument<Boolean>("enabled") ?: false
                if (enabled) {
                    meetingSession?.audioVideo?.startLocalVideo()
                } else {
                    meetingSession?.audioVideo?.stopLocalVideo()
                }
                result.success(null)
            }
            "switchCamera" -> {
                meetingSession?.audioVideo?.switchCamera()
                applyLocalMirror()
                result.success(null)
            }
            "setSpeakerEnabled" -> {
                val enabled = call.argument<Boolean>("enabled") ?: false
                setSpeaker(enabled)
                result.success(null)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun joinMeeting(meetingMap: Map<String, Any>, attendeeMap: Map<String, Any>, videoEnabled: Boolean) {
        leaveMeeting(resetAudio = false)

        val meetingId = meetingMap["MeetingId"] as? String ?: throw IllegalArgumentException("Missing MeetingId")
        val externalMeetingId = meetingMap["ExternalMeetingId"] as? String ?: ""
        val mediaPlacementMap = meetingMap["MediaPlacement"] as? Map<*, *> ?: throw IllegalArgumentException("Missing MediaPlacement")

        val audioHostUrl = mediaPlacementMap["AudioHostUrl"] as? String ?: ""
        val audioFallbackUrl = mediaPlacementMap["AudioFallbackUrl"] as? String ?: ""
        val signalingUrl = mediaPlacementMap["SignalingUrl"] as? String ?: ""
        val turnControlUrl = mediaPlacementMap["TurnControlUrl"] as? String ?: ""
        val eventIngestionUrl = mediaPlacementMap["EventIngestionUrl"] as? String

        val mediaPlacement = MediaPlacement(
            AudioFallbackUrl = audioFallbackUrl,
            AudioHostUrl = audioHostUrl,
            SignalingUrl = signalingUrl,
            TurnControlUrl = turnControlUrl
        )

        val meeting = Meeting(
            ExternalMeetingId = externalMeetingId,
            MediaPlacement = mediaPlacement,
            MediaRegion = meetingMap["MediaRegion"] as? String ?: "us-east-1",
            MeetingId = meetingId
        )

        val attendeeId = attendeeMap["AttendeeId"] as? String ?: throw IllegalArgumentException("Missing AttendeeId")
        val externalUserId = attendeeMap["ExternalUserId"] as? String ?: ""
        val joinToken = attendeeMap["JoinToken"] as? String ?: throw IllegalArgumentException("Missing JoinToken")

        localAttendeeId = attendeeId
        desiredMuted = false

        val attendee = Attendee(
            AttendeeId = attendeeId,
            ExternalUserId = externalUserId,
            JoinToken = joinToken
        )

        val configuration = MeetingSessionConfiguration(
            CreateMeetingResponse(meeting),
            CreateAttendeeResponse(attendee)
        )

        val session = DefaultMeetingSession(
            configuration,
            ConsoleLogger(),
            context
        )

        meetingSession = session

        session.audioVideo.addAudioVideoObserver(this)
        session.audioVideo.addVideoTileObserver(this)
        session.audioVideo.addRealtimeObserver(this)

        session.audioVideo.start()
        // Enable RECEIVING remote video. This is the master switch  it calls
        // VideoClient.setReceiving(true) under the hood. Without it the video
        // client never accepts ANY inbound video frames, so onRemoteVideoSourceAvailable
        // can fire and we can call updateVideoSourceSubscriptions all we want, but
        // onVideoTileAdded for the remote tile never fires and the peer is stuck on
        // "Waiting for video". updateVideoSourceSubscriptions only PICKS sources; it
        // does NOT flip the receiving flag. Audio + local self-view work without this,
        // which is exactly the symptom we were seeing. Always start it (cheap no-op for
        // audio-only calls  no remote video sources will be advertised).
        session.audioVideo.startRemoteVideo()
        if (videoEnabled) {
            session.audioVideo.startLocalVideo()
        }
    }

    private fun leaveMeeting(resetAudio: Boolean = true) {
        mainHandler.removeCallbacks(audioResetRunnable)
        meetingSession?.let { session ->
            try {
                session.audioVideo.realtimeLocalUnmute()
            } catch (e: Exception) {
                Log.w(TAG, "Unable to unmute before leaving meeting", e)
            }
            try {
                session.audioVideo.stopLocalVideo()
            } catch (e: Exception) {
                Log.w(TAG, "Unable to stop local video before leaving meeting", e)
            }
            try {
                session.audioVideo.stopRemoteVideo()
            } catch (e: Exception) {
                Log.w(TAG, "Unable to stop remote video before leaving meeting", e)
            }
            try {
                session.audioVideo.stop()
            } catch (e: Exception) {
                Log.w(TAG, "Unable to stop meeting audio/video", e)
            } finally {
                session.audioVideo.removeAudioVideoObserver(this)
                session.audioVideo.removeVideoTileObserver(this)
                session.audioVideo.removeRealtimeObserver(this)
            }
        }
        meetingSession = null
        localTileId = null
        remoteTileId = null
        localAttendeeId = null
        desiredMuted = false
        if (resetAudio) {
            resetAndroidAudioAfterCall()
            mainHandler.postDelayed(audioResetRunnable, AUDIO_RESET_DELAY_MS)
        }
    }

    private fun applyLocalMute(audioVideo: AudioVideoFacade, muted: Boolean) {
        if (muted) {
            audioVideo.realtimeLocalMute()
            sendEvent("localMuted")
        } else {
            audioVideo.realtimeLocalUnmute()
            sendEvent("localUnmuted")
        }
    }

    private fun resetAndroidAudioAfterCall() {
        val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as? AudioManager ?: return
        try {
            audioManager.isMicrophoneMute = false
        } catch (e: Exception) {
            Log.w(TAG, "Unable to clear microphone mute after call", e)
        }
        try {
            audioManager.isSpeakerphoneOn = false
        } catch (e: Exception) {
            Log.w(TAG, "Unable to clear speaker route after call", e)
        }
        try {
            if (audioManager.mode == AudioManager.MODE_IN_COMMUNICATION) {
                audioManager.mode = AudioManager.MODE_NORMAL
            }
        } catch (e: Exception) {
            Log.w(TAG, "Unable to reset audio mode after call", e)
        }
    }

    private fun setSpeaker(enabled: Boolean) {
        val audioVideo = meetingSession?.audioVideo ?: return
        val devices = audioVideo.listAudioDevices()
        val targetType = if (enabled) MediaDeviceType.AUDIO_BUILTIN_SPEAKER else MediaDeviceType.AUDIO_HANDSET
        val device = devices.find { it.type == targetType }
        if (device != null) {
            audioVideo.chooseAudioDevice(device)
        } else if (!enabled) {
            // fallback to wired headset if handset not found
            val fallbackDevice = devices.find { it.type == MediaDeviceType.AUDIO_WIRED_HEADSET }
            if (fallbackDevice != null) {
                audioVideo.chooseAudioDevice(fallbackDevice)
            }
        }
    }

    /**
     * Mirror the local self-view ONLY when the front camera is active, the way
     * users expect a selfie preview to behave. Back camera is never mirrored.
     * `mirror` is a render-only flag on the local DefaultVideoRenderView, so the
     * stream sent to the remote peer is unaffected  they always see us
     * correctly oriented, and captured/published frames are not flipped.
     */
    private fun applyLocalMirror() {
        val isFront = meetingSession?.audioVideo?.getActiveCamera()?.type ==
            MediaDeviceType.VIDEO_FRONT_CAMERA
        activeLocalView?.getVideoRenderView()?.mirror = isFront
    }

    fun registerVideoView(view: ChimeVideoPlatformView, isLocal: Boolean) {
        mainHandler.post {
            if (isLocal) {
                activeLocalView = view
                localTileId?.let { tileId ->
                    meetingSession?.audioVideo?.bindVideoView(view.getVideoRenderView(), tileId)
                }
                applyLocalMirror()
            } else {
                activeRemoteView = view
                remoteTileId?.let { tileId ->
                    meetingSession?.audioVideo?.bindVideoView(view.getVideoRenderView(), tileId)
                }
            }
        }
    }

    fun unregisterVideoView(view: ChimeVideoPlatformView) {
        mainHandler.post {
            if (view == activeLocalView) {
                activeLocalView = null
                localTileId?.let { tileId ->
                    meetingSession?.audioVideo?.unbindVideoView(tileId)
                }
            } else if (view == activeRemoteView) {
                activeRemoteView = null
                remoteTileId?.let { tileId ->
                    meetingSession?.audioVideo?.unbindVideoView(tileId)
                }
            }
        }
    }

    private fun sendEvent(eventName: String, data: Map<String, Any> = emptyMap()) {
        mainHandler.post {
            eventSink?.success(mapOf("event" to eventName) + data)
        }
    }

    // StreamHandler implementations
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    // AudioVideoObserver implementations
    override fun onAudioSessionStarted(reconnecting: Boolean) {
        meetingSession?.audioVideo?.let { audioVideo ->
            applyLocalMute(audioVideo, desiredMuted)
        }
        sendEvent(if (reconnecting) "reconnecting" else "connected")
    }

    override fun onAudioSessionStartedConnecting(reconnecting: Boolean) {
        sendEvent("connecting")
    }

    override fun onAudioSessionStopped(sessionStatus: MeetingSessionStatus) {
        sendEvent("disconnected")
    }

    override fun onAudioSessionCancelledReconnect() {
        sendEvent("disconnected")
    }

    override fun onConnectionRecovered() {
        sendEvent("connected")
    }

    override fun onAudioSessionDropped() {
        sendEvent("failed")
    }

    override fun onConnectionBecamePoor() {}
    override fun onVideoSessionStartedConnecting() {}
    override fun onVideoSessionStarted(sessionStatus: MeetingSessionStatus) {}
    override fun onVideoSessionStopped(sessionStatus: MeetingSessionStatus) {}
    override fun onCameraSendAvailabilityUpdated(available: Boolean) {}

    // Chime SDK 0.21.0 does NOT auto-deliver remote video. When a peer turns on
    // their camera the SDK advertises the source here, and we must explicitly
    // subscribe or onVideoTileAdded never fires for the remote tile  which is
    // exactly why remote video never rendered. Subscribe to every advertised
    // source at high priority/resolution (1:1 calls only ever have one).
    override fun onRemoteVideoSourceAvailable(sources: List<RemoteVideoSource>) {
        val audioVideo = meetingSession?.audioVideo ?: return
        if (sources.isEmpty()) return
        val subscriptions = sources.associateWith {
            VideoSubscriptionConfiguration(VideoPriority.Highest, VideoResolution.High)
        }
        audioVideo.updateVideoSourceSubscriptions(subscriptions, emptyArray())
    }

    override fun onRemoteVideoSourceUnavailable(sources: List<RemoteVideoSource>) {
        val audioVideo = meetingSession?.audioVideo ?: return
        if (sources.isEmpty()) return
        audioVideo.updateVideoSourceSubscriptions(emptyMap(), sources.toTypedArray())
    }

    // VideoTileObserver implementations
    override fun onVideoTileAdded(tileState: VideoTileState) {
        if (tileState.isLocalTile) {
            localTileId = tileState.tileId
            sendEvent("localVideoEnabled")
            activeLocalView?.let { view ->
                meetingSession?.audioVideo?.bindVideoView(view.getVideoRenderView(), tileState.tileId)
            }
            applyLocalMirror()
        } else {
            remoteTileId = tileState.tileId
            sendEvent("remoteVideoEnabled")
            activeRemoteView?.let { view ->
                meetingSession?.audioVideo?.bindVideoView(view.getVideoRenderView(), tileState.tileId)
            }
        }
    }

    override fun onVideoTileRemoved(tileState: VideoTileState) {
        if (tileState.isLocalTile) {
            localTileId = null
            sendEvent("localVideoDisabled")
            meetingSession?.audioVideo?.unbindVideoView(tileState.tileId)
        } else {
            remoteTileId = null
            sendEvent("remoteVideoDisabled")
            meetingSession?.audioVideo?.unbindVideoView(tileState.tileId)
        }
    }

    override fun onVideoTilePaused(tileState: VideoTileState) {}
    override fun onVideoTileResumed(tileState: VideoTileState) {}
    override fun onVideoTileSizeChanged(tileState: VideoTileState) {}

    // RealtimeObserver implementations
    override fun onAttendeesJoined(attendeeInfo: Array<AttendeeInfo>) {
        for (info in attendeeInfo) {
            if (info.attendeeId != localAttendeeId) {
                sendEvent("remoteJoined")
            }
        }
    }

    override fun onAttendeesLeft(attendeeInfo: Array<AttendeeInfo>) {
        for (info in attendeeInfo) {
            if (info.attendeeId != localAttendeeId) {
                sendEvent("remoteLeft")
            }
        }
    }

    override fun onAttendeesDropped(attendeeInfo: Array<AttendeeInfo>) {
        for (info in attendeeInfo) {
            if (info.attendeeId != localAttendeeId) {
                sendEvent("remoteLeft")
            }
        }
    }

    override fun onAttendeesMuted(attendeeInfo: Array<AttendeeInfo>) {
        for (info in attendeeInfo) {
            if (info.attendeeId == localAttendeeId) {
                desiredMuted = true
                sendEvent("localMuted")
            }
        }
    }

    override fun onAttendeesUnmuted(attendeeInfo: Array<AttendeeInfo>) {
        for (info in attendeeInfo) {
            if (info.attendeeId == localAttendeeId) {
                desiredMuted = false
                sendEvent("localUnmuted")
            }
        }
    }
    override fun onSignalStrengthChanged(signalUpdates: Array<SignalUpdate>) {}
    override fun onVolumeChanged(volumeUpdates: Array<VolumeUpdate>) {}
}
